odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+        """
+        super(CupyTensorSpace, self).__init__(shape, dtype)
+        if self.dtype.char not in self.available_dtypes():
+            raise ValueError('`dtype` {!r} not supported'.format(dtype))
+
+        if device is None:
+            self.__device = cupy.cuda.get_device_id()
+        else:
+            self.__device = int(device)
+
+        dist = kwargs.pop('dist', None)
+        norm = kwargs.pop('norm', None)
+        inner = kwargs.pop('inner', None)
+        weighting = kwargs.pop('weighting', None)
+        exponent = kwargs.pop('exponent', 2.0)
+
+        # Check validity of option combination (3 or 4 out of 4 must be None)
+        if sum(x is None for x in (dist, norm, inner, weighting)) < 3:
+            raise ValueError('invalid combination of options `weighting`, '
+                             '`dist`, `norm` and `inner`')
+        if any(x is not None for x in (dist, norm, inner)) and exponent != 2.0:
+            raise ValueError('`exponent` cannot be used together with '
+                             '`dist`, `norm` and `inner`')
+
+        # Set the weighting
+        if weighting is not None:
+            if isinstance(weighting, Weighting):
+                if weighting.impl != 'cupy':
+                    raise ValueError("`weighting.impl` must be 'cupy', "
+                                     '`got {!r}'.format(weighting.impl))
+                if weighting.exponent != exponent:
+                    raise ValueError('`weighting.exponent` conflicts with '
+                                     '`exponent`: {} != {}'
+                                     ''.format(weighting.exponent, exponent))
+                self.__weighting = weighting
+            else:
+                self.__weighting = _weighting(weighting, exponent)
+
+            # Check (afterwards) that the weighting input was sane
+            if isinstance(self.weighting, CupyTensorSpaceArrayWeighting):
+                if not np.can_cast(self.weighting.array.dtype, self.dtype):
+                    raise ValueError(
+                        'cannot cast from `weighting` data type {} to '
+                        'the space `dtype` {}'
+                        ''.format(dtype_str(self.weighting.array.dtype),
+                                  dtype_str(self.dtype)))
+                if self.weighting.array.shape != self.shape:
+                    raise ValueError('array-like weights must have same '
+                                     'shape {} as this space, got {}'
+                                     ''.format(self.shape,
+                                               self.weighting.array.shape))
+
+        elif dist is not None:
+            self.__weighting = CupyTensorSpaceCustomDist(dist)
+        elif norm is not None:
+            self.__weighting = CupyTensorSpaceCustomNorm(norm)
+        elif inner is not None:
+            self.__weighting = CupyTensorSpaceCustomInner(inner)
+        else:  # all None -> no weighing
```

Comment:
What happened to our no weighting vs const weighting debate? I.e. in several cases we expect stuff like this to be weighted:

```
space(weighting=n/10.0)  # "randomly" gives no weighting for n=10
```

I guess here it matters less than for the discretized spaces, but in them it really maters.

```
spc = odl.uniform_discr(0, 5, n)  # 'randomly' not weighted for n=5
```

this causes problems downstream for e.g. `RayTransform` which has special behaviour for non-weighted spaces.

Reply:
I'd still need a convincing argument why weighting with constant 1 would be fundamentally different from no weighting. I currently see it just as an optimization that scraps a multiplication with 1.

Reply:
Is this not a very weird behaviour:

```
>>> spc = odl.uniform_discr([0, 0], [9.999, 9.999], [10, 10])
>>> rt = odl.tomo.RayTransform(spc, odl.tomo.parallel_beam_geometry(spc))
>>> rt.range
uniform_discr([  0.    , -14.1407], [  3.1416,  14.1407], (45, 31))
```

vs

```
>>> spc = odl.uniform_discr([0, 0], [9.999, 9.999], [10, 10])
>>> rt = odl.tomo.RayTransform(spc, odl.tomo.parallel_beam_geometry(spc))
>>> rt.range
uniform_discr([  0.    , -14.1421], [  3.1416,  14.1421], (45, 31), weighting=1.0)
```

I mean, wat? this *has to* be considered a rather severe bug, no?

Reply:
That seems quite wrong. 

But part of the problem seems to me that we need to jump through hoops currently to make the adjoint correct for weighted vs. unweighted spaces. Doing it via something like #1177 will solve the issue very elegantly by keeping all weighting stuff completely outside the operator definitions.
And when such a system is in place, no weighting and weighting with 1 will be *exactly* equivalent.

Reply:
Well this has nothing to do with the adjoint (in fact we don't even need the adjoint to have this problem), the problem here is that we want "the same weighting scheme" on the range as on the domain, but if we dont distinguish no weighting and 1.0 weight, we can't do that.

Reply:
Oh right, I was totally forgetting about the fact that we infer the range. Okay now I got it. Hm. We do have an optional `range` parameter for the ray transform, don't we? I tend to think that the behavior should be 

- as currently when we have a `TensorSpace` with `cell_volume` weighting,
- no weighting in any other case.

To make it easier to change this, we could have a `asweighted()` method on `TensorSpace` where a new space of the same type but with new weighting can be created.

Reply:
Well it's not quite obvious to me how we would figure out if we're using cell_volume weighting 1.0 or no weighting unless we expose this flag somehow. We previously did it by exposing thetype of the weighting, but I guess we could fix this by exposing the argument used to create the weighting (somehow), i.e. `space.weighting_type` returns a string `'const'` or w/e. Does that sound reasonable?

I feel adding yet another `as_...` should hopefully not be needed right now, but I guess in the long run.